kimchi/src/snarky/traits.rs

```diff
@@ -0,0 +1,161 @@
+    cvar::CVar,
+};
+
+/// A snarky type is a type that can be used in a circuit.
+/// It references an equivalent "out-of-circuit" type that one can use outside of the circuit.
+/// (For example, to construct private or public inputs, or a public output, to the circuit.)
+pub trait SnarkyType<F>: Sized
+where
+    F: PrimeField,
+{
+    /// Some 'out-of-circuit' data, which is carried as part of Self.
+    /// This data isn't encoded as CVars in the circuit, since the data may be large (e.g. a sparse merkle tree),
+    /// or may only be used by witness computations / for debugging.
+    type Auxiliary;
+
+    /// The equivalent "out-of-circuit" type.
+    /// For example, the [super::boolean::Boolean] snarky type has an out-of-circuit type of [bool].
+    type OutOfCircuit;
+
+    /// The number of field elements that this type takes.
+    const SIZE_IN_FIELD_ELEMENTS: usize;
+
+    /// Returns the circuit variables (and auxiliary data) behind this type.
+    fn to_cvars(&self) -> (Vec<CVar<F>>, Self::Auxiliary);
+
+    /// Creates a new instance of this type from the given circuit variables (And some auxiliary data).
+    fn from_cvars_unsafe(cvars: Vec<CVar<F>>, aux: Self::Auxiliary) -> Self;
+
+    /// Checks that the circuit variables behind this type are valid.
+    /// For some definition of valid.
+    /// For example, a Boolean snarky type would check that the field element representing it is either 0 or 1.
+    /// The function does this by adding constraints to your constraint system.
+    fn check(&self, cs: &mut RunState<F>);
+
+    /// The "default" value of [Self::Auxiliary].
+    /// This is passed to [Self::from_cvars_unsafe] when we are not generating a witness,
+    /// since we have no candidate value to get the auxiliary data from.
+    /// Note that we use an explicit value here rather than Auxiliary: Default,
+    /// since the default value for the type may not match the default value we actually want to pass!
+    fn constraint_system_auxiliary() -> Self::Auxiliary;
+
+    /// Converts an out-of-circuit value
+    fn value_to_field_elements(value: &Self::OutOfCircuit) -> (Vec<F>, Self::Auxiliary);
+
+    fn value_of_field_elements(fields: Vec<F>, aux: Self::Auxiliary) -> Self::OutOfCircuit;
+
+    //
+    // new functions that might help us with generics?
+    //
+
+    fn compute<FUNC>(cs: &mut RunState<F>, loc: String, to_compute_value: FUNC) -> Self
+    where
+        FUNC: Fn(&dyn WitnessGeneration<F>) -> Self::OutOfCircuit,
+    {
+        cs.compute(loc, to_compute_value)
+    }
+
+    fn read<G>(&self, g: G) -> Self::OutOfCircuit
+    where
+        G: WitnessGeneration<F>,
+    {
+        let (cvars, aux) = self.to_cvars();
+        let values = cvars.iter().map(|cvar| g.read_var(cvar)).collect();
+        Self::value_of_field_elements(values, aux)
+    }
+}
+
+//
+// Auto traits
+//
+
+impl<F> SnarkyType<F> for ()
+where
+    F: PrimeField,
+{
+    type Auxiliary = ();
```

Comment:
Can we provide a default for this associated type? Or is that not supported

Reply:
oh yeah I think we can do this!